### PR TITLE
[WIP] Actually implement this

### PR DESCRIPTION
**Pull request by dignifiedquire**
The dragons still live here. So beware of them. 

## Things left to do 

- [x] [external] release libp2p-record https://github.com/libp2p/js-libp2p-record/pull/1
- [x] [external] Update libp2p peer events https://github.com/libp2p/js-libp2p/issues/74
- [x] [external] Make peerbook in libp2p be the canoncial list of all seen peers https://github.com/ipfs/js-libp2p-ipfs-nodejs/pull/91
- [x] ~~[external] Add latency recording to peer-book~~ (moved to https://github.com/libp2p/js-peer-book/issues/34)
- [x] ~~[external] Add close method to interface-connection~~ (moved to https://github.com/libp2p/js-libp2p-swarm/issues/205)
- [x] [external] Integrate into libp2p https://github.com/libp2p/js-libp2p/pull/86
- [ ] [external] Integrate into js-ipfs https://github.com/ipfs/js-ipfs/pull/822
- [ ] [external] Integrate into js-ipfs-bitswap https://github.com/ipfs/js-ipfs-bitswap/pull/142
- [ ] Fix all todos
- [x] ~~Figure out if go uses lp messages or not~~, It does, need to implement it 
- [x] Implement bootstrap
- [x] Implement providers interface
- [x] Implement peer query interface
- `dht_test.go`
  - [x] get-set
  - [x] providers
  - [x] bootstrap
  - [x] layered get
  - [x] findPeer
  - [x] ~~findPeersConnectedToPeer~~ (removing, doesn't work in go and will be implemented at a later point)
  - [x] connect collision
  - [x] bad proto messages
  - [x] findPeer query
  - [x] findClosestPeers
- `ext_test.go`
  - [ ] get failures
  - [ ] not found
  - [ ] less than K responses
- [x] Enough tests for good coverage stats
  - [x] Statements   : Current: 86% - Goal: 90% fine enough :)
  - [ ] Branches     : Current: 67% - Goal: 90% - will improve at a later point
  - [x] Functions    : Current: 99% - Goal: 100% fine enough :)
  - [x] Lines        : Current: 86% - Goal: 90% fine enough :)
- [ ] Write interop tests with go
  - [ ] in js-ipfs
  - [x] encoded messages
- [x] add `start` and `stop`
- [ ] Documentation
  - [x] API
  - [ ] Overview with graph and things

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+language: node_js
+node_js:
+  - 4
```
Comment (diasdavid) [inline .travis.yml]: 6 and stable too

File: README.md [hunk 2]
```diff
@@ -1,13 +1,53 @@
...
+[![Build Status](https://travis-ci.org/ipfs/js-libp2p-dht.svg?style=flat-square)](https://travis-ci.org/ipfs/js-libp2p-dht)
+[![Coverage Status](https://coveralls.io/repos/github/ipfs/js-libp2p-dht/badge.svg?branch=master)](https://coveralls.io/github/ipfs/js-libp2p-dht?branch=master)
+[![Dependency Status](https://david-dm.org/ipfs/js-libp2p-dht.svg?style=flat-square)](https://david-dm.org/ipfs/js-libp2p-dht)
+[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat-square)](https://github.com/feross/standard)
+[![standard-readme compliant](https://img.shields.io/badge/standard--readme-OK-green.svg?style=flat-square)](https://github.com/RichardLitt/standard-readme)
+![](https://img.shields.io/badge/npm-%3E%3D3.0.0-orange.svg?style=flat-square)
+![](https://img.shields.io/badge/Node.js-%3E%3D4.0.0-orange.svg?style=flat-square)
 
-> JavaScript implementation of the DHT for libp2p
+> JavaScript implementation of the Kademlia DHT for libp2p, based on [go-kad-dht](https://github.com/libp2p/go-libp2p-kad-dht).
```
Comment (diasdavid) [inline README.md]: s/go-kad-dht/go-libp2p-kad-dht

**Comment by diasdavid** [comment]
@dignifiedquire seeing an error in occasionally:

```
  1) DHT find peer query:

      Uncaught AssertionError: expected 1 to deeply equal 0
      + expected - actual

      -1
      +0

      at utils.sortClosestPeers (test/index.spec.js:240:51)

```

File: src/index.js [hunk 3]
```diff
@@ -0,0 +1,565 @@
...
+  /**
+   * Start the bootstrap process. This means running a number of queries every interval requesting random data.
+   * This is done to keep the dht healthy over time.
+   *
+   * @param {number} [queries=1] - how many queries to run per period
+   * @param {number} [period=300000] - how often to run the the bootstrap process, in milliseconds (5min)
+   * @param {number} [maxTimeout=10000] - how long to wait for the the bootstrap query to run, in milliseconds (10s)
+   * @returns {void}
+   */
+  bootstrapStart (queries, period, maxTimeout) {
```
Comment (diasdavid) [inline src/index.js]: This is, in fact, a discovery service that DHT provides and so as any discovery service it needs to be passed into the discovery mechanisms

**Comment by diasdavid** [comment]
Needs CR:

- [ ] https://github.com/libp2p/js-libp2p-identify/pull/40
- [ ] https://github.com/libp2p/js-libp2p-swarm/pull/218
- [ ] https://github.com/ipfs/js-libp2p-ipfs-nodejs/pull/100
- [ ] https://github.com/libp2p/js-libp2p/pull/86

**Comment by diasdavid** [comment]
@dignifiedquire can you elaborate on the work behind (and importance of)

-  [ ] Fix all todos
- `ext_test.go`
  - [ ] get failures
  - [ ] not found
  - [ ] less than K responses

File: src/index.js (hunk 3 above)
Comment (dignifiedquire) [inline src/index.js]: hmm

File: .travis.yml (hunk 1 above)
Comment (diasdavid) [inline .travis.yml]: update CI

File: README.md (hunk 2 above)
Comment (diasdavid) [inline README.md]: update

File: README.md [hunk 4]
```diff
@@ -1,13 +1,53 @@
...
+- [API](#api)
+- [Contribute](#contribute)
+- [License](#license)
+
+## Install
+
+### npm
+
+```sh
+> npm i libp2p-dht
```
Comment (diasdavid) [inline README.md]: s/libp2p-dht/libp2p-kad-dht/g

**Comment by dignifiedquire** [comment]
> @dignifiedquire can you elaborate on the work behind (and importance of)

These are nice to haves to check for specific failure behaviour but not crucial for shipping.

File: README.md [hunk 5]
```diff
@@ -1,13 +1,53 @@
...
+```
+
+### Use in Node.js
+
+```js
+const KadDHT = require('libp2p-dht')
```
Comment (diasdavid) [inline README.md]: s/libp2p-dht/libp2p-kad-dht/g

File: README.md [hunk 6]
```diff
@@ -1,13 +1,53 @@
...
+```
+
+## API
+
+See https://libp2p.github.io/js-libp2p-dht
```
Comment (diasdavid) [inline README.md]: s/libp2p-dht/libp2p-kad-dht/g

**Pull request merged**